docs/proposals/202105256-cluster-class-and-managed-topologies.md

```diff
@@ -0,0 +1,507 @@
+      // +optional
+      RolloutAfter *metav1.Time `json:"rolloutAfter,omitempty"`
+
+      // The information for the Control plane of the cluster.
+      ControlPlane ControlPlane `json:"controlPlane"`
+
+      // Worker encapsulates the different constructs that form the worker nodes
+      // for the cluster.
+      // +optional
+      Worker *Worker `json:"worker,omitempty"`
+    }
+    ```
+1.  The `ControlPlane` object contains the parameters for the control plane nodes of the managed topology.
+    ```golang
+    // ControlPlane specifies the parameters for the control plane nodes in the cluster.
+    type ControlPlane struct {
+     Metadata ObjectMeta `json:"metadata,omitempty"`
+
+     // The number of control plane nodes.
+     Replicas int `json:"replicas"`
+    }
+    ```
+1.  The `Worker` object represents the sets of worker nodes of the managed topology.
+
+    **Note**: In this proposal, a set of worker nodes is handled by a MachineDeployment object. In the future, this can be extended to include Machine Pools as another backing mechanism for managing worker node sets.
+    ```golang
+    // Worker represents the different sets of worker nodes in the cluster.
+    type Worker struct {
+      Deployments []ManagedDeployment `json:”deployments,omitempty”`
+    }
+    ```
+1.  The `ManagedDeployment` object represents a single set of worker nodes of the managed topology.
+    ```golang
+    // ManagedDeployment specifies the different parameters for a set of worker nodes in the cluster.
+    type ManagedDeployment struct {
+      Metadata ObjectMeta `json:"metadata,omitempty"`
+
+      // Class is the name of the DeploymentClass used to create the set of worker nodes.
+      // This should match one of the deployment classes defined in the ClusterClass object
+      // mentioned in the `Cluster.Spec.Class` field.
+      Class string `json:"class"`
+
+      // Name refers to the name of the set of worker nodes.
+      Name string `json:"name"`
+
+      // The number of worker nodes belonging to this set.
+      // If not specified, the cluster autoscaler for CAPI would be responsible
+      // for maintaining a min/max number of replicas.
+      // +optional
+      Replicas *int `json:"replicas,omitempty"`
+    }
+    ```
+
+#### Validations
+##### ClusterClass
+- For object creation:
+  - `spec.worker.deployments[i].class` field must be unique within a ClusterClass.
+- For object updates:
+  - `spec.worker.deployments` supports adding new deployment classes.
+
+##### Cluster
+- For object creation:
+  - `spec.topology` and `spec.controlPlaneRef` cannot be simultaneously set.
+  - If `spec.topology.class` is set, `spec.topology.version` cannot be empty and must be a valid semver.
+
+- For object updates:
+  - If spec.topology.class is in use, it cannot be unset or modified.
+  - spec.topology.version cannot be unset and must be a valid semver, if being updated.
+  - `spec.topology.version` cannot be downgraded.
```

Comment:
How do we support rollbacks for scenarios where machineDeployments get stuck during an upgrade?

Reply:
Added a **Future goal** to implement an upgrade/rollback strategy.

Reply:
Similar to the above comment around version upgrades.